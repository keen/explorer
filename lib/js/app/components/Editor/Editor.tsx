import React, { FC } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Alert, Button } from '@keen.io/ui-core';

import { EditorActions, CreatorContainer } from './Editor.styles';

import EditorNavigation from '../EditorNavigation';

import Creator from '../Creator';

import RunQuery, { runQueryLabel } from '../RunQuery';
import QueryVisualization from '../QueryVisualization';
import VisualizationPlaceholder from '../VisualizationPlaceholder';

import {
  getQueryResults,
  getQueryPerformState,
  getError,
} from '../../modules/queries';
import { setViewMode } from '../../modules/app';

type Props = {
  /** Query definition */
  query: Record<string, any>;
  /** Query update handler */
  onUpdateQuery: (query: Record<string, any>) => void;
  /** Run query event handler */
  onRunQuery: () => void;
};

const Editor: FC<Props> = ({ query, onRunQuery, onUpdateQuery }) => {
  const dispatch = useDispatch();
  const queryResults = useSelector(getQueryResults);
  const runQueryError = useSelector(getError);
  const isQueryLoading = useSelector(getQueryPerformState);

  return (
    <div>
      <EditorNavigation query={query} />
      <Button
        onClick={() => {
          dispatch(setViewMode('browser'));
        }}
      >
        Back to list
      </Button>
      <section>
        {queryResults ? (
          <QueryVisualization query={query} queryResults={queryResults} />
        ) : (
          <VisualizationPlaceholder isLoading={isQueryLoading} />
        )}
        {runQueryError && <Alert type="error">{runQueryError.body}</Alert>}
      </section>
      <CreatorContainer>
        <Creator onUpdateQuery={onUpdateQuery} />
      </CreatorContainer>
      <section>
        <EditorActions>
          <RunQuery isLoading={isQueryLoading} onClick={onRunQuery}>
            {runQueryLabel(query)}
          </RunQuery>
        </EditorActions>
      </section>
    </div>
  );
};

export default Editor;
