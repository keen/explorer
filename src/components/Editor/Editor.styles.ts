import styled from 'styled-components';
import { transparentize } from 'polished';
import { colors } from '@keen.io/colors';
import { BACKGROUND_MAIN } from '../../constants';

export const EditorActions = styled.div`
  display: flex;
  padding: 0 20px 20px 20px;
  background: ${BACKGROUND_MAIN};
`;

export const ButtonWrapper = styled.div`
  margin-left: 10px;
`;

export const CreatorContainer = styled.div`
  padding: 20px;
  background: ${BACKGROUND_MAIN};
`;

export const Card = styled.div`
  background: ${colors.white[500]};
  box-shadow: 0 2px 4px 0 ${transparentize(0.85, colors.black[500])};
`;
