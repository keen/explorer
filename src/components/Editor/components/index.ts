import SectionTabs from './SectionTabs';
import Visualization from './Visualization';
import Creator from './Creator';

export { SectionTabs, Creator, Visualization };
