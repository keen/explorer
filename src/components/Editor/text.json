{
  "clearButton": "Clear",
  "extractToEmailButton": "Extract to Email"
}
