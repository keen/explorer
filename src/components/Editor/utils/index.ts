import { showEmailExtraction } from './showEmailExtraction';

export { showEmailExtraction };
