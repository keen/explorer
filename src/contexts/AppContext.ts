import React from 'react';
import { PubSub } from '@keen.io/pubsub';

import { DatavizSettings } from '../types';

const AppContext = React.createContext<{
  keenAnalysis: any;
  notificationPubSub: PubSub;
  modalContainer: string;
  upgradeSubscriptionUrl?: string;
  datavizSettings: DatavizSettings;
}>({
  keenAnalysis: null,
  modalContainer: null,
  notificationPubSub: null,
  upgradeSubscriptionUrl: null,
  datavizSettings: {},
});

export default AppContext;
