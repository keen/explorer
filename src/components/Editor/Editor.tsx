import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { Button } from '@keen.io/ui-core';

import {
  EditorActions,
  Card,
  ButtonWrapper,
  CreatorContainer,
} from './Editor.styles';

import { showEmailExtraction } from './utils';

import EditorNavigation from '../EditorNavigation';
import Creator from '../Creator';

import RunQuery, { runQueryLabel } from '../RunQuery';
import EditorVisualization from '../EditorVisualization';
import VisualizationPlaceholder from '../VisualizationPlaceholder';
import QueryLimitReached from '../QueryLimitReached';

import {
  extractToEmail,
  getQueryResults,
  getQueryPerformState,
  getQueryLimitReached,
  setQuerySettings,
} from '../../modules/queries';
import {
  clearQuery,
  updateChartSettings as updateSettings,
} from '../../modules/app';

type Props = {
  /** Query definition */
  query: Record<string, any>;
  /** Run query event handler */
  onRunQuery: () => void;
  /** Save query event handler */
  onSaveQuery: () => void;
  /** Optional upgrade subscription url */
  upgradeSubscriptionUrl?: string;
};

const Editor: FC<Props> = ({
  query,
  upgradeSubscriptionUrl,
  onRunQuery,
  onSaveQuery,
}) => {
  const dispatch = useDispatch();
  const { t } = useTranslation();

  const queryResults = useSelector(getQueryResults);
  const isQueryLoading = useSelector(getQueryPerformState);
  const isQueryLimitReached = useSelector(getQueryLimitReached);

  const updateQuery = useCallback((query: Record<string, any>) => {
    dispatch(setQuerySettings(query));
  }, []);

  const updateChartSettings = useCallback(
    (chartSettings: Record<string, any>) => {
      dispatch(updateSettings(chartSettings));
    },
    []
  );

  return (
    <div id="editor">
      <EditorNavigation onSaveQuery={onSaveQuery} />
      <section>
        {isQueryLimitReached && (
          <QueryLimitReached upgradeSubscriptionUrl={upgradeSubscriptionUrl} />
        )}
        <Card>
          {queryResults && !isQueryLimitReached && (
            <EditorVisualization query={query} queryResults={queryResults} />
          )}
          {!queryResults && !isQueryLimitReached && (
            <VisualizationPlaceholder isLoading={isQueryLoading} />
          )}
        </Card>
      </section>
      <CreatorContainer>
        <Creator
          onUpdateQuery={updateQuery}
          onUpdateChartSettings={updateChartSettings}
        />
      </CreatorContainer>
      <EditorActions>
        <RunQuery isLoading={isQueryLoading} onClick={() => onRunQuery()}>
          {t(runQueryLabel(query))}
        </RunQuery>
        {showEmailExtraction(query) && (
          <ButtonWrapper data-testid="email-extraction">
            <Button
              variant="success"
              size="large"
              onClick={() => dispatch(extractToEmail())}
            >
              {t('editor.extract_to_email_button')}
            </Button>
          </ButtonWrapper>
        )}
        <ButtonWrapper>
          <Button
            onClick={() => dispatch(clearQuery())}
            style="outline"
            variant="success"
            size="large"
          >
            {t('editor.clear_query_button')}
          </Button>
        </ButtonWrapper>
      </EditorActions>
    </div>
  );
};

export default Editor;
