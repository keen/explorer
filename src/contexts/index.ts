import AppContext from './AppContext';

export { AppContext };
