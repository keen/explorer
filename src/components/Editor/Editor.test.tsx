/* eslint-disable @typescript-eslint/camelcase */
import React from 'react';
import { Provider } from 'react-redux';
import { render as rtlRender, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';

import { AppContext } from '../../contexts';

import Editor from './Editor';

jest.mock('@keen.io/query-creator');

const render = (storeState: any = {}, overProps: any = {}) => {
  const mockStore = configureStore([]);
  const state = {
    savedQuery: {
      name: '',
      displayName: '',
      cached: false,
      refreshRate: 0,
      exists: false,
      tags: [],
    },
    app: {
      querySettingsModal: {
        visible: true,
      },
    },
    queries: {
      results: null,
      isLoading: false,
      queriesExecution: {
        limitReached: false,
      },
    },
    ...storeState,
  };

  const store = mockStore({ ...state });

  const props = {
    query: {},
    onUpdateQuery: jest.fn(),
    onRunQuery: jest.fn(),
    onSaveQuery: jest.fn(),
    ...overProps,
  };

  const wrapper = rtlRender(
    <AppContext.Provider value={{ keenAnalysis: { config: {} } } as any}>
      <Provider store={store}>
        <Editor {...props} />
      </Provider>
    </AppContext.Provider>
  );

  return {
    store,
    props,
    wrapper,
  };
};

test('allows user to reset query settings', () => {
  const {
    wrapper: { getByText },
    store,
  } = render();

  const clearButton = getByText('editor.clear_query_button');
  fireEvent.click(clearButton);

  expect(store.getActions()).toMatchInlineSnapshot(`
    Array [
      Object {
        "payload": undefined,
        "type": "@app/QUERY_EDITOR_MOUNTED",
      },
      Object {
        "payload": undefined,
        "type": "@app/CLEAR_QUERY",
      },
    ]
  `);
});

test('allows user to run query', () => {
  const {
    wrapper: { getByText },
    props,
  } = render();

  const clearButton = getByText('editor.run_query_button');
  fireEvent.click(clearButton);

  expect(props.onRunQuery).toHaveBeenCalled();
});

test('do not renders email extraction button', () => {
  const {
    wrapper: { queryByText },
  } = render();

  expect(queryByText('editor.extract_to_email_button')).not.toBeInTheDocument();
});

test('allows user to perform extraction to email', () => {
  const overProps = {
    query: {
      analysis_type: 'extraction',
    },
  };
  const {
    wrapper: { getByText },
    store,
  } = render({}, overProps);

  const button = getByText('editor.extract_to_email_button');
  fireEvent.click(button);

  expect(store.getActions()).toMatchInlineSnapshot(`
    Array [
      Object {
        "payload": undefined,
        "type": "@app/QUERY_EDITOR_MOUNTED",
      },
      Object {
        "payload": undefined,
        "type": "@queries/EXTRACT_TO_EMAIL",
      },
    ]
  `);
});

test('renders notice about queries execution limit exceed', () => {
  const storeState = {
    queries: {
      results: null,
      isLoading: false,
      queriesExecution: {
        limitReached: true,
      },
    },
  };

  const {
    wrapper: { getByTestId },
  } = render(storeState);

  expect(getByTestId('query-execution-limit')).toBeInTheDocument();
});
