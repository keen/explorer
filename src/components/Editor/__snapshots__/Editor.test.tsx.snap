// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows user to customize widget subtitle 1`] = `
Array [
  Object {
    "payload": Object {
      "widgetSettings": Object {
        "card": Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 0,
          "enabled": true,
          "hasShadow": true,
        },
        "legend": Object {
          "alignment": "left",
          "enabled": true,
          "layout": "horizontal",
          "position": "top",
        },
        "subtitle": Object {
          "content": "@subtitle",
        },
        "title": Object {
          "content": "",
        },
      },
    },
    "type": "@app/UPDATE_WIDGET_SETTINGS",
  },
]
`;

exports[`allows user to customize widget title 1`] = `
Array [
  Object {
    "payload": Object {
      "widgetSettings": Object {
        "card": Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 0,
          "enabled": true,
          "hasShadow": true,
        },
        "legend": Object {
          "alignment": "left",
          "enabled": true,
          "layout": "horizontal",
          "position": "top",
        },
        "subtitle": Object {
          "content": "",
        },
        "title": Object {
          "content": "@title",
        },
      },
    },
    "type": "@app/UPDATE_WIDGET_SETTINGS",
  },
]
`;
